04-onion-routing.md

```diff
@@ -2,35 +2,40 @@
 
 ## Overview
 
-This document describes the construction of an onion routed packet that is used to route a payment from a _origin node_ to a _final node_, over a number of intermediate nodes, called _hops_.
+This document describes the construction of an onion routed packet that is
+used to route a payment from an _origin node_ to a _final node_. The packet
+is routed through a number of intermediate nodes which are referred to as _hops_.
 
 The routing schema is based on the
 [Sphinx](http://www.cypherpunks.ca/~iang/pubs/Sphinx_Oakland09.pdf)
-construction, and is extended with a per-hop payload.
+construction and is extended with a per-hop payload.
 
 Intermediate nodes forwarding the message can verify the integrity of
-the packet, and can learn about which node they should forward the
-packet to. They cannot learn about which other nodes, besides their
-predecessor or successor, are part of this route, nor can they learn
-the length of the route and their position within it. The packet is
-obfuscated at each hop, so that a network level attacker cannot
-associate packets belonging to the same route, i.e., packets belonging
-to a route do not share any identifying information. Notice that this
-does not preclude the possibility to associate packets by performing a
-traffic analysis.
-
-The route is constructed by the sender, which knows a public key of
-each intermediate node. Knowing the intermediate node's public key
-allows the sender to create a shared secret using ECDH for each
-intermediate node, including the final recipient. The shared secret is then
-used to generate a _pseudo-random stream_ of bytes to obfuscate the
-packet, and a number of _keys_ used to encrypt the payload and compute
-HMACs ensuring integrity at each hop.
+the packet and can learn which node they should forward the
+packet to. They cannot learn which other nodes, besides their
+predecessor or successor, are part of the packet's route; nor can they learn
+the length of the route or their position within it. The packet is
+obfuscated at each hop, to ensure that a network level attacker cannot
+associate packets belonging to the same route, i.e. packets belonging
+to the same route do not share any identifying information. Notice that this
+does not preclude the possibility of packet association by an attacker
+via traffic analysis.
+
+The route is constructed by the sender (origin node), which knows the public keys of
+each intermediate node. Knowing each intermediate node's public key
+allows the sender to create a shared secret (using ECDH) for each
+intermediate node, including the final recipient node. The shared secret is then
+used to generate a _pseudo-random stream_ of bytes, which is used to obfuscate the
+packet, and a number of _keys_, which are used to encrypt the payload and compute
+HMACs, which are used to ensure the integrity of the packet at each hop.
```

Comment:
The recursive whiches at the end read a bit awkwardly, but I can take a stab at that when I get around to a full edit on 5.

Reply:
Noted, let's do in another pass?

Reply:
fixed